kj-rs/maybe.rs

```diff
@@ -0,0 +1,246 @@
+                match value {
+                    None => Maybe {
+                        is_set: false,
+                        some: MaybeUninit::zeroed(),
+                    },
+                    Some(val) => Maybe {
+                        is_set: true,
+                        some: MaybeUninit::new(val),
+                    }
+                }
+            }
+        }
+    };
+    ($ty:ty, $($tail:ty),+) => {
+        impl_maybe_item_for_primitive!($ty);
+        impl_maybe_item_for_primitive!($($tail),*);
+    };
+}
+
+impl_maybe_item_for_has_niche!(crate::Own<T>, &T, &mut T, *const T, *mut T);
+impl_maybe_item_for_primitive!(
+    u8, u16, u32, u64, u128, i8, i16, i32, i64, i128, usize, isize, bool
+);
+
+pub(crate) mod repr {
+    use super::MaybeItem;
+    use static_assertions::assert_eq_size;
+    use std::fmt::{Debug, Display};
+    use std::mem::MaybeUninit;
+
+    /// A [`Maybe`] represents bindings to the `kj::Maybe` class.
+    /// It is an optional type, but represented using a struct, for alignment with kj.
+    ///
+    /// # Layout
+    /// In kj, `Maybe` has 3 specializations, one without niche value optimization, and
+    /// two with it. In order to maintain an identical layout in Rust, we include an associated type
+    /// in the [`MaybeItem`] trait, which determines the discriminant of the `Maybe<T: MaybeItem>`.
+    ///
+    /// ## Niche Value Optimization
+    /// This discriminant is used in tandem with the [`crate::maybe::HasNiche`] to implement
+    /// [`MaybeItem`] properly for values which have a niche, which uses a discriminant of [`()`],
+    /// the unit type. All other types use [`bool`].
+    #[repr(C)]
+    pub struct Maybe<T: MaybeItem> {
+        pub(super) is_set: T::Discriminant,
+        pub(super) some: MaybeUninit<T>,
+    }
+
+    assert_eq_size!(Maybe<isize>, [usize; 2]);
+    assert_eq_size!(Maybe<&isize>, usize);
+    assert_eq_size!(Maybe<crate::Own<isize>>, [usize; 2]);
+    assert_eq_size!(Maybe<*const isize>, usize);
+
+    impl<T: MaybeItem> Maybe<T> {
+        /// # Safety
+        /// This function shouldn't be used except by macro generation.
+        pub unsafe fn is_set(&self) -> T::Discriminant {
+            self.is_set
+        }
+
+        /// # Safety
+        /// This function shouldn't be used except by macro generation.
+        pub unsafe fn from_parts_unchecked(
+            is_set: T::Discriminant,
+            some: MaybeUninit<T>,
+        ) -> Maybe<T> {
+            Maybe { is_set, some }
+        }
+
+        pub fn is_some(&self) -> bool {
+            T::is_some(self)
+        }
+
+        pub fn is_none(&self) -> bool {
+            T::is_none(self)
+        }
+    }
+
+    impl<T: MaybeItem> From<Maybe<T>> for Option<T> {
+        fn from(value: Maybe<T>) -> Self {
+            if value.is_some() {
+                // We can't move out of value so we copy it and forget it in
+                // order to perform a "manual" move out of value
```

Comment:
Is this because `value` isn't a mutable type? Is that a limitation of the From trait?